docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,186 @@
+# Comparing UniFFI with Diplomat
+
+[Diplomat](https://github.com/rust-diplomat/diplomat/) and [UniFFI](https://github.com/mozilla/uniffi-rs/)
+are both tools which expose a rust implemented API over an FFI.
+At face value, these tools are solving the exact same problem, but their approach
+is significantly different.
+
+This document attempts to describe these different approaches and discuss the pros and cons of each.
+It's not going to try and declare one better than the other, but instead just note how they differ.
+If you are reading this hoping to find an answer to "what one should I use?", then that's easy -
+each tool currently supports a unique set of foreign language bindings, so the tool you should
+use is the one that supports the languages you care about!
+
+(There may even be a future where these 2 tools converge - that seems like a lot of work, but
+might also provide a large payoff - more on this later)
+
+Disclaimer: This document was written by one of the UniFFI developers, who has never used
+diplomat in anger. Please feel free to open PRs if anything here misrepresents diplomat.
+
+# The type systems
+
+The key different between these 2 tools is the "type system". While both are exposing Rust
+code (which obviously comes with its own type system), the foreign bindings need to know
+lots of details about all the types expressed by the tool.
+
+For the sake of this document, we will use the term "type universe" to define the set of
+all types known by each of the tools. Both of these tools build their own "type universe" then
+use that to generate both Rust code and foreign bindings.
+
+## UniFFI's type universe
+UniFFI's model is to parse an external ffi description from a `.udl` file which describes the
+entire "type universe". This type universe is then used to generate both the Rust scaffolding
+(on disk as a `.rs` file) and the foreign bindings.
+
+**What's good about this** is that the entire type system is known when generating both the rust code
+and the foreign binding.
```

Comment:
Actually, the point I had in mind was simply that the foreign bindings need to know the names and types of struct elements so it can recreate the same struct on the other side. But I've tried to capture that in the new version.